File: Study Plan/Top Interview 150/190 Reverse Bits/ReverseBits.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    190. Reverse Bits
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/reverse-bits/solutions/4384305/clean-solution-in-c-beats-96-50-of-users-in-runtime/

    - Runtime 16 ms (Beats 96.34% of users with C#)
    - Memory 24.56 MB (Beats 18.50% of users with C#)
*/
public class Solution
{
    public uint reverseBits(uint n)
    {
        var original = Convert.ToString(n, 2).PadLeft(32, '0').ToCharArray();
        Array.Reverse(original);
        var reversed = new string(original);
        
        return Convert.ToUInt32(reversed, 2);
    }
}

File: Study Plan/Top Interview 150/0190 Reverse Bits/ReverseBits.java
/*  
    Problem
    --------------------------------------------------------------------------------
    190. Reverse Bits
*/

/*  
    Solution
    --------------------------------------------------------------------------------

    - Runtime 5 ms (Beats 5.28% of users with Java)
    - Memory 42.17 MB (Beats 5.68% of users with Java)
*/

class Solution {
    // you need treat n as an unsigned value
    public int reverseBits(int n) {
        // Returns a string representation of the integer argument as an
        // unsigned integer in base 2.
        String binaryString = Integer.toBinaryString(n);
        binaryString = String.format("%32s", binaryString).replace(' ', '0');
        StringBuilder reversed = new StringBuilder(binaryString).reverse();
        long unsigned = Long.parseLong(reversed.toString(), 2);
        return (int)unsigned;
    }
}